raft membership: fix promotion/demotion workflow

abronan opened the pull request:
Simply rolls back the role on `node demote` if this fails. There is opportunity for a refactoring here as we can add a check upstream and only selectively update node Spec fields, but not sure if really necessary.

/cc @aaronlehmann @tonistiigi 

Signed-off-by: Alexandre Beslic <alexandre.beslic@gmail.com>

aaronlehmann commented:
LGTM

tonistiigi commented:
Isn't there a race in here depending how much time it takes for the restore to run. Also the `Node` directly  picks up spec changes so they will still get a change to worker and try to renew certificate. Not sure if it will cause a conflict atm. with demotion but it does leak invalid role.

abronan commented:
You're right, we should probably add the check on demote upstream to the Spec changes so that it does not trigger the cert renewal. I'll rework on the PR to move the `CanRemoveMember` check followed by the Removal. This way this shouldn't trigger the cert renewal and this shouldn't race.

I just wonder what should we do in this case to remain consistent with the Spec update? Should we update everything but the role or nothing on the Spec depending on the demote failure?

/cc @aluzzardi ^^^

abronan commented:
@aaronlehmann @tonistiigi PTAL, I refactored `UpdateNode` and also fixed the scaling down to one manager in the process. I exported `Cluster` management from raft to separate raw Remove from Remove with quorum failure checks and updated the tests to reflect the demotion to one manager.

@tonistiigi I'm going to give it a try myself, but would be awesome if you can check the behavior on the engine.

aaronlehmann commented on manager/controlapi/node.go:
```diff
@@ -189,14 +189,18 @@ func (s *Server) UpdateNode(ctx context.Context, request *api.UpdateNodeRequest)
 	)
+
+	s.raft.MembershipLock.Lock()
+	defer s.raft.MembershipLock.Unlock()
```
Let's avoid this if at all possible. Can we expose a RemoveMemberIfAllowed function in raft.go?

aaronlehmann commented on manager/controlapi/node.go:
```diff
@@ -208,6 +212,19 @@ func (s *Server) UpdateNode(ctx context.Context, request *api.UpdateNodeRequest)
+				return grpc.Errorf(codes.FailedPrecondition, "attempting to demote a manager while this might result in a loss of quorum")
+			}
+
+			if err := s.raft.RemoveMember(ctx, raftID); err != nil {
```
Why is this inside an Update callback? It will hold the memory store lock for potentially quite a long time.

tonistiigi commented on manager/controlapi/node.go:
```diff
@@ -208,6 +212,19 @@ func (s *Server) UpdateNode(ctx context.Context, request *api.UpdateNodeRequest)
 			}
 		}
 
+		member := s.raft.Cluster.GetMemberByNodeID(request.NodeID)
```
Possible nil dereference here. This should probably be inside the if clause.

abronan commented:
Thanks, I'll take care of the comments. Just found a remaining issue where you try to demote yourself as a member, the node would be listed as `UNKNOWN` and removed but still with the Manager role. Have to fix that and add a few more test cases for `UpdateNode`.

tonistiigi commented:
@abronan I also have some problems with demoting leader. I get a `context canceled` error and the promoting same node doesn't seem to work(at least for the first time).

```
root@5e0ac9a640fc:/go/src/github.com/docker/docker# docker swarm init
Swarm initialized: current node (3w2gvjjuczre2e7gvaxeqa77p) is now a manager.
root@5e0ac9a640fc:/go/src/github.com/docker/docker# docker -H 0.0.0.0:4010 swarm join --listen-addr 0.0.0.0:2378 0.0.0.0:2377
This node joined a Swarm as a worker.
root@5e0ac9a640fc:/go/src/github.com/docker/docker# docker node ls
ID                           HOSTNAME      MEMBERSHIP  STATUS  AVAILABILITY  MANAGER STATUS
3w2gvjjuczre2e7gvaxeqa77p *  5e0ac9a640fc  Accepted    Ready   Active        Leader
dy5kefi0a5ne9xlc0p5s8wuex    5e0ac9a640fc  Accepted    Ready   Active
root@5e0ac9a640fc:/go/src/github.com/docker/docker# docker node promote dy5kefi0a5ne9xlc0p5s8wuex
Node dy5kefi0a5ne9xlc0p5s8wuex promoted to a manager in the swarm.
root@5e0ac9a640fc:/go/src/github.com/docker/docker# docker -H 0.0.0.0:4010 node demote 3w2gvjjuczre2e7gvaxeqa77p
Error response from daemon: rpc error: code = 1 desc = context canceled
root@5e0ac9a640fc:/go/src/github.com/docker/docker# docker -H 0.0.0.0:4010  node ls
ID                           HOSTNAME      MEMBERSHIP  STATUS  AVAILABILITY  MANAGER STATUS
3w2gvjjuczre2e7gvaxeqa77p    5e0ac9a640fc  Accepted    Ready   Active
dy5kefi0a5ne9xlc0p5s8wuex *  5e0ac9a640fc  Accepted    Ready   Active        Leader
root@5e0ac9a640fc:/go/src/github.com/docker/docker# docker -H 0.0.0.0:4010  node promote 3w2gvjjuczre2e7gvaxeqa77p
Node 3w2gvjjuczre2e7gvaxeqa77p promoted to a manager in the swarm.
root@5e0ac9a640fc:/go/src/github.com/docker/docker# docker -H 0.0.0.0:4010  node ls
ID                           HOSTNAME      MEMBERSHIP  STATUS  AVAILABILITY  MANAGER STATUS
3w2gvjjuczre2e7gvaxeqa77p    5e0ac9a640fc  Accepted    Ready   Active
dy5kefi0a5ne9xlc0p5s8wuex *  5e0ac9a640fc  Accepted    Ready   Active        Leader
```

abronan commented:
@aaronlehmann @tonistiigi I was actually wondering if this makes any sense to demote yourself as the Leader. My argument was that it's like shooting yourself in the foot as the node will demote itself and then the clusterapi would be unavailable until it elects a new Leader (it can be `5s` but the `ElectionTick` is configurable and can be high for some users in high-latency networks).

I understand the argument for flexibility to allow users to demote whatever the node they want but anything that results in cluster downtime from direct user action should be avoided imho. And this risks to cause a lot of confusion too.

This does not feel unreasonable to me to disallow demoting the Leader as we can support scaling down to one node and we can't demote the last node anyway (which in this case, will be the Leader).

/cc @aluzzardi

aaronlehmann commented:
IMHO we should allow demoting the leader. The impact is only a few seconds of time where you can't change the cluster state while you wait for a new election. For many lab/dev setups this is a non-event, and it's a legitimate use case where you want to change the manager topology. I think if we don't allow this, people will complain - and the risk of allowing it is pretty small.

tonistiigi commented:
@abronan I think not allowing to demote yourself is fine(currently it ends with "connection is closing" message), but demoting a leader should be allowed. In the example above I use another node to demote the leader.

stevvooe commented on manager/state/raft/raft.go:
```diff
@@ -74,7 +74,7 @@ const (
 type Node struct {
 	raft.Node
-	cluster *membership.Cluster
+	Cluster *membership.Cluster
```
Why are we exporting this field?

stevvooe commented on manager/state/raft/raft.go:
```diff
@@ -121,7 +121,7 @@ type Node struct {
 	stopMu sync.RWMutex
 	// used for membership management checks
-	membershipLock sync.Mutex
+	MembershipLock sync.Mutex
```
Seem like this should be encapsulated...

abronan commented:
@tonistiigi I fixed the "connection is closing" so you can now even demote yourself (from Leader or Follower) and this works fine. Currently double checking some other things but should push an update soon.

Still not convinced with allowing to demote the Leader as I think it encourages bad actions on the cluster but I'll lean to the majority and leave it possible to do it. If we consider it dangerous in the future, it's just a simple check to add so I'm fine with allowing it for now.

LK4D4 commented:
@abronan cool! do you want to run me integration tests?

aaronlehmann commented on manager/state/raft/membership/cluster.go:
```diff
@@ -166,18 +166,14 @@ func (c *Cluster) ValidateConfigurationChange(cc raftpb.ConfChange) error {
 	members := c.Members()
 
-	nmembers := 0
+	nmembers := 1
```
How about `nmembers := len(members) + 1`? And probably rename it too, since it's no longer the number of members.

aaronlehmann replied:
Better yet, just use `len(members) + 1` in the one place we use it. Problem solved.

LK4D4 commented:
integration tests are passed

stevvooe commented on manager/controlapi/node.go:
```diff
@@ -199,13 +206,26 @@ func (s *Server) UpdateNode(ctx context.Context, request *api.UpdateNodeRequest)
+			}
+
+			// Quorum safeguard
+			if !s.raft.CanRemoveMember(member.RaftID) {
```
Presumably, this is protected by the updateLock to some degree, such that if we demote two nodes concurrently, one will fail?

abronan replied:
Correct.

abronan commented:
@aaronlehmann @LK4D4 @tonistiigi PTAL took care of most comments, added a simple test case in `controlapi` for the scaling down to one Manager that should be complementary to the integration tests.

stevvooe commented on manager/controlapi/node.go:
```diff
@@ -220,14 +240,19 @@ func (s *Server) UpdateNode(ctx context.Context, request *api.UpdateNodeRequest)
-				break
-			}
+
+				node.Spec = initSpec
```
If this node is killed in the middle of this operation, won't this `initSpec` get lost, since the change is propagated by the above transaction?

aaronlehmann replied:
It looks like this is a rollback that happens if we tried to remove the manager from the raft cluster, but failed. I think you're right that this rollback won't happen if the leader dies between the two transactions.

I'm thinking maybe we shouldn't do this rollback, for a few reasons:

- The client may have changed other things in the spec. Here we're discarding all those other changes because the demotion failed.
- It's a spec, so it should be under control of the user. We shouldn't be changing things based on actual state.
- The corner case @stevvooe points out here with node failure.

abronan replied:
> The client may have changed other things in the spec. Here we're discarding all those other changes because the demotion failed.

I've asked for that in a previous comment, but imho I think we should not allow the Spec change if the role change fails. This gives a sense that the Spec change is `incomplete`, which can be confusing to the user that might expect the demotion to take effect as much as all the other changes from the Spec. And the issue here is that we don't try (yet) to reconcile the raft memberlist from what is in the Spec.

> The corner case @stevvooe points out here with node failure.

The corner case is definitely possible, but it's better than the current behavior which does not fix the role on `Remove` failure and leaves the node stuck to role changes.

I'm not against reverting to the either doing the rollback for the role only (Spec being partially applied), or removing this entirely. Even though we can leave this here as a best effort mechanism to rollback in case of a failure on `Removal`.

The real solution would be to make the demotion happen in the Spec and then try to `Remove` the node from raft until it succeeds to match the behavior of other Spec changes, but outside of the call to `UpdateNode`.

stevvooe replied:
@abronan I think the notion that we have to rollback the spec is probably incorrect. The spec reflects the desired state. If the desired state cannot be achieved, we report on that and keep trying.

Is the issue that we are querying the spec for the current role, rather than gating it through a node field?

aaronlehmann replied:
> I've asked for that in a previous comment, but imho I think we should not allow the Spec change if the role change fails.

I would be okay with that in principle, but the code isn't implementing that quite correctly. Instead of not allowing it to change, it's doing a rollback on failure. But the rollback is not atomic, and if something else changes the spec in between the two transactions, their changes will be lost.

> I'm not against reverting to the either doing the rollback for the role only (Spec being partially applied), or removing this entirely.

I think what would be best for now is to not roll anything back, and leave a TODO about reattempting reconcilation of the desired state.

> Is the issue that we are querying the spec for the current role, rather than gating it through a node field?

I think this is absolutely the issue. If we had the notion of "observed role", it would solve this. Spec would be entirely under the user's control, and "observed role" would only update on a successful promotion/demotion.

aaronlehmann commented:
LGTM

Let's merge this (it's only the uncontroversial parts) and open a separate issue/PR about handling failed demotion more gracefully.

LK4D4 commented:
Yeah, let's merge it.
ping @stevvooe

abronan commented:
#1090 is tracking the next steps. Let's keep it opened for now.
